### PR TITLE
Fixed connection leak in basic and digest authenticators if 401 response has content

### PR DESCRIPTION
**Pull request by agherardi**
This is a clone of https://github.com/jersey/jersey/pull/3623

### CODE REVIEW

File: core-client/src/main/java/org/glassfish/jersey/client/authentication/DigestAuthenticator.java [hunk 1]
```diff
@@ -349,6 +352,25 @@ private String randomBytes(final int nbBytes) {
         return bytesToHex(bytes);
     }
 
+    private static void discardInputAndClose(InputStream is) {
```
Comment (jansupol) [inline core-client/src/main/java/org/glassfish/jersey/client/authentication/DigestAuthenticator.java]: Do we really need to consume the stream before it is closed?

Comment (agherardi) [reply core-client/src/main/java/org/glassfish/jersey/client/authentication/DigestAuthenticator.java]: Consuming the stream makes it possible for the connection to be reused (unless the server closes the connection, of course).

File: core-client/src/main/java/org/glassfish/jersey/client/authentication/BasicAuthenticator.java [hunk 2]
```diff
@@ -96,11 +99,33 @@ public boolean filterResponseAndAuthenticate(ClientRequestContext request, Clien
...
+                    discardInputAndClose(response.getEntityStream());
+                }
                 throw new ResponseAuthenticationException(null, LocalizationMessages.AUTHENTICATION_CREDENTIALS_MISSING_BASIC());
             }
 
             return HttpAuthenticationFilter.repeatRequest(request, response, calculateAuthentication(credentials));
         }
         return false;
     }
+
```
Comment (jansupol) [inline core-client/src/main/java/org/glassfish/jersey/client/authentication/BasicAuthenticator.java]: This discardInputAndClose is once again in HttpAuthenticationFilter in PR3860.  Can you put it to the AuthenticationUtil as discussed in PR 3862?

**Comment by agherardi** [comment]
Hi - Any updates?

**Comment by jansupol** [comment]
There is class AuthenticationUtil #discardInputAndClose already from the previous PR. Can you please rebase and resolve conflicts?

**Comment by agherardi** [comment]
Conflict resolved. There is a test error. Looking into it.

**Comment by agherardi** [comment]
The test issue is caused by a missing "@GET" in front of this https://github.com/eclipse-ee4j/jersey/blob/master/connectors/apache-connector/src/test/java/org/glassfish/jersey/apache/connector/AuthTest.java#L157 test method, due to a merge conflict with https://github.com/eclipse-ee4j/jersey/pull/3858/files.

I re-added the "@GET" with this merge request. Otherwise, please feel free to back out https://github.com/eclipse-ee4j/jersey/pull/3858, I'll re-submit with the fixed test.

**Comment by agherardi** [comment]
Thanks. This merge request should be ready to be merged now.

**Comment by agherardi** [comment]
The IP validation check is failing. Apparently, a couple of merge commits don't have either the author or the signed-off line in the commit message.

I'm assuming the changes can still be merged since the "important" commit has both author and signed-off. If that's not the case, please let me know - I'll open a new pull request. Thanks.

**Comment by jansupol** [comment]
There seems to be some discrepancy in the sign-off/commit users alessandro.gherardi@yahoo.com vs alessandro.gherardi@schneider-electric.com. Not sure where it took the latter user. Can you please better create a new PR? Thanks

**Comment by agherardi** [comment]
Changed re-submitted as https://github.com/eclipse-ee4j/jersey/pull/3878 . Closing this request.

**Pull request closed without merging**